### DISCUSSION TITLE
Can't see traces in the webUI if callback_handler passed to LLM init constructor

### DISCUSSION THREAD

**Discussion by gabrielfior** [langfuse/langfuse#1864] [Support]
I went through the Langfuse docs regarding [Langchain integration](https://langfuse.com/docs/integrations/langchain/example-python), and I understand that passing the callback to the functions `invoke` as well as `chain.run` and `predict` sends traces to Langfuse. However, what I would like to have is to pass the callback handlers directly to the LLM init method (see code below) and all subsequent traces be logged to Langfuse.

```
llm = ChatOpenAI(
            model="gpt-4-turbo-2024-04-09",
            api_key=XXXX,
            temperature=0.0,
            callbacks=[CallbackHandler(
            secret_key=XXX,
            public_key=XXX,
            host=XXX)
            ],
        )
```

Please note the following
-> If I execute methods like `llm.invoke`, I see the traces in the UI
-> If I pass this llm to an CrewAI Agent, which is then called within a CrewAI crew, no traces are displayed in the webUI.

I expected this to work out-of-the-box.
Guidance much appreciated!

**Comment by saibhaskerraju** [comment]
The problem is with CrewAI callback. 

do the following

1. add the callback handler to LLM init

2. add the same callback instance to agent callback .
3. Do not add anything to crewai run or anywhere else

hope i helps. i am doing the same

**Reply by sbhadana** [reply]
its not working, with crewai. could you please share some working langfuse CallbackHandler?
Thanks

**Reply by saibhaskerraju** [reply]
` from textwrap import dedent
from crewai import Crew, Agent, Task, Process
from crewai_tools import WebsiteSearchTool
from langchain_openai import AzureChatOpenAI
from langchain_community.tools import DuckDuckGoSearchRun
import os
from dotenv import load_dotenv
load_dotenv()
from langfuse.callback import CallbackHandler

langfuse_handler = CallbackHandler(
                                  )

langfuse_handler.auth_check()

model = AzureChatOpenAI(
 deployment_name="gpt-4-preview",
 model_name="gpt-4-preview",
 callbacks=[langfuse_handler]
)

researcher = Agent(
 role='Senior Research Analyst',
 llm = model,
 backstory=dedent(
    """This agent excels in internet research, keeping tabs on regulatory updates, technological advancements,
    and industry trends to inform audit planning and strategies with external insights."""
   ),
 goal=dedent(
    """To provide comprehensive, up-to-date information from the internet, enriching audit plans with the latest
    external insights and best practices across industries."""
   ),
 tools=[DuckDuckGoSearchRun()],
 verbose=True,
 callbacks=[langfuse_handler]
)

writer = Agent(
 role='Content Writer',
 llm = model,
 backstory='Write engaging articles on AI discoveries',
 goal='Write engaging articles on AI discoveries',
 verbose=True,
 callbacks=[langfuse_handler] 
)
research_task = Task(
 description='Identify breakthrough AI technologies',
 expected_output='Identify breakthrough AI technologies',
 agent=researcher
)
write_article_task = Task(
 description='Draft an article on the latest AI technologies',
 expected_output='Draft an article on the latest AI technologies',
 agent=writer
)
my_crew = Crew(
 agents=[researcher, writer],
 tasks=[research_task, write_article_task],
 process=Process.sequential,
 full_output=True,
 verbose=True,
)
result = my_crew.kickoff()
print(result)
`
